core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala

```diff
@@ -1129,6 +1129,36 @@ class RDDSuite extends SparkFunSuite with SharedSparkContext {
     }.collect()
   }
 
+  test("SPARK-23496: order of input partitions can result in severe skew in coalesce") {
+    val numInputPartitions = 100
+    val numCoalescedPartitions = 50
+    val locations = Array("locA", "locB")
+
+    val inputRDD = sc.makeRDD(Range(0, numInputPartitions).toArray[Int], numInputPartitions)
+    assert(inputRDD.getNumPartitions == numInputPartitions)
+
+    val locationPrefRDD = new LocationPrefRDD(inputRDD, { (p: Partition) =>
+      if (p.index < numCoalescedPartitions) {
+        Seq(locations(0))
+      } else {
+        Seq(locations(1))
+      }
+    })
+    val coalescedRDD = new CoalescedRDD(locationPrefRDD, numCoalescedPartitions)
+
+    val numPartsPerLocation = coalescedRDD
+      .getPartitions
+      .map(coalescedRDD.getPreferredLocations(_).head)
+      .groupBy(identity)
+      .mapValues(_.size)
+
+    // Without the fix these would be:
+    //  numPartsPerLocation(locations(0)) == numCoalescedPartitions - 1
+    //  numPartsPerLocation(locations(1)) == 1
+    assert(numPartsPerLocation(locations(0)) > 0.4 * numCoalescedPartitions)
```

Comment:
How confident are we on the assert condition to be true? How is the fraction `0.4` chosen?

Reply:
Added comment about flakiness & fixed seed.